### ISSUE TITLE
Configure Rotation Origin

### ISSUE THREAD

**Issue by jcampbell05** [unidoc/unidoc#159]
We have code to scale down PDFs to A4 documents and center them, as part of this we want to output a portrait PDF. If the input PDF was landscape this means they end up much smaller so we want to make sure they are 90 degrees rotate so they benefit from having as much space as possible.

In the current API it rotates from some sort of origin the library defaults to, we looked through the code and it looks like the only way to control this is to hack it by doing some translates before rotating.

I noticed you have a TODO to control the origin which we would be interested in as we want to be able to control it so we can simply rotate by 90 degrees and center it (The centering code completely breaks at the moment as it assumes the origin is in the middle.

**Comment by gunnsth** [comment]
Can you post an example code snippet and an example PDF?

Its worth keeping in mind that rotation in PDF can be done in two ways:
1. Page rotation flag: multiple of 90 degrees.
2. Content rotation

The rotation is performed about the upper left corner.  So in your case the content would probably disappear if rotated by 90 degrees counterclockwise.

Translation is needed to counter the effect of the rotation about the upper left corner.
See for example our example:
https://github.com/unidoc/unidoc-examples/blob/master/pdf/pages/pdf_rotate_flatten.go

which sets the page rotation flag to 0, but accounts for the rotation in the content stream (flattening).

Yes, in the TODO you mentioned, the idea is to be allow to specify the pivot point around which the rotation is done, for example in the middle of the page which would probably do what you want without any translation.  Typically rotating about the upper left corner and then just translating afterwards is fine, so I would recommend doing that.

**Comment by gunnsth** [comment]
@jcampbell05 I believe #376 (just merged into v3 today) addresses this. Would be great if you can check and confirm whether it works.

It now rotates about the center by default.  We believe this is usually the desired behavior.  We will consider adding custom rotation origins only if there is a clear need/use case.

**Comment by jcampbell05** [comment]
@gunnsth sure, this isn't a priority on our side right now but when I have some free time I'll test it out. But moving the origin to the center will be useful to us. 

Will this be a breaking change for code built around the previous behaviour ?

**Comment by gunnsth** [comment]
@jcampbell05 Yes, there could be breaking changes in terms of the output.  Thus, it would be great if you can comment on whether this solves the original issue.  It will be the default in next stable major version v3.0.0.

We avoid breaking output changes generally although there can be exceptions such as this one.  In those cases it goes hand in hand with major version upgrades.
